nvisii: Fix NDDS export and other fixes

mintar opened the pull request:
This is another big round of fixes and improvements to the nvisii data generation script. The json format should now be compatible with NDDS and the old training script (except for the `_camera_settings.json` and `_object_settings.json` files, which still have to be generated manually). The PR is split into logical commits with their own commit messages, so please review commit-by-commit, but here are the highlights:

* Document the dataset format in the README
* Change camera coordinate system to NDDS/OpenCV
* Fix 2d bounding box computation when height != width
* Update `train.py` so that it understands the `projected_cuboid` field with 9 elements instead of a separate `projected_cuboid_centroid` field
* Add --focal-length parameter
* Compute the visibility fraction by default and add `px_count_all` and `px_count_visib` fields so that the dataset can be converted to the BOP format later
* Fix handling of class names with underscores
* more NDDS compatibility fixes

TontonTremblay commented:
Sorry I have been stormed with other projects. I am looking at it now.

TontonTremblay commented:
So I downloaded your update. It looks sharp as always. I have some comments. Looking at the readme, I do not see how to enable the script to generate jsons like ndds. Could we add `--ndds` or maybe it is just the default version and we can remove the train2.py script. I would be fine with that. 

Is there some doc about how to use debug_json_ros_node.py and its utility? 

I had the code to not do percentage by default since it can be quite costly. But for the toy example it does not impact perf. But on some other scenes it has been pretty intense, maybe we could move it to args. Not a big deal for now. Ok I just noticed the argument, you beaten me to the punch. 

Looking again, you did not make this nvisii script export similar data to ndds, you made the OG train file compatible with this ndds data? If it is the case, we should update the readme to say so. Thoughts? 


Thank you so much with this update. To me it looks clean. I will accept it.

mintar commented:
Thanks for the quick review! Here are some answers to your comments. I've also added a new PR #236 that addresses them.

> Looking at the readme, I do not see how to enable the script to generate jsons like ndds. Could we add --ndds or maybe it is just the default version and we can remove the train2.py script. I would be fine with that.
> [...]
> Looking again, you did not make this nvisii script export similar data to ndds, you made the OG train file compatible with this ndds data? If it is the case, we should update the readme to say so. Thoughts?

After merging #226, #234 and #236, the generated jsons are fully compatible with the old training script. The only significant change is that the NDDS json files have 8 elements in `projected_cuboid` and 1 element in `projected_cuboid_centroid`, whereas the json files generated by the nvisii script have 9 elements in `projected_cuboid` and no `projected_cuboid_centroid` field. I could have changed that too, but instead I made the `train.py` script detect this automatically in 9b509fa. I've added a comment to the README in #236.

I agree that the `train2/train.py` script is no longer necessary just to use the data generated by the nvisii script, but before deleting it I'd like to check if there are any other parts in there that should be merged into `train.py`.

> Is there some doc about how to use debug_json_ros_node.py and its utility?

Good point. I've added some doc directly to the source file in #236. It was mainly useful for me while debugging all the different transformations in the json file, so now it's not really necessary any more. Perhaps it's useful to look at for people who want to understand how the different transformations play together. We could also remove it if you wish, I don't care.

> I had the code to not do percentage by default since it can be quite costly. But for the toy example it does not impact perf. But on some other scenes it has been pretty intense, maybe we could move it to args. Not a big deal for now. Ok I just noticed the argument, you beaten me to the punch.

Yes, I've noticed that too after I made the change. Of course it depends on how many objects are visible in the scene. I've profiled it with 50-75 target objects and 25 distractors, and enabling the visibility fraction caused a 50% overhead. I've changed the default back to `False` so that it doesn't compute the visibility fraction by default in a #236.

> Thank you so much with this update. To me it looks clean. I will accept it.

Thanks! It was really a lot of work, but I think it's worth it. The PBR images generated by the nvisii script are just beautiful.
